noir-projects/aztec-nr/aztec/src/oracle/keys.nr

```diff
@@ -0,0 +1,76 @@
+use dep::protocol_types::{
+  address::{
+    AztecAddress,
+    PartialAddress,
+    PublicKeysHash,
+  }, 
+  constants::{
+    GENERATOR_INDEX__PUBLIC_KEYS_HASH,
+    GENERATOR_INDEX__CONTRACT_ADDRESS_V1,
+  },
+  grumpkin_point::GrumpkinPoint,
+};
+
+use crate::hash::poseidon2_hash;
+
+#[oracle(getPublicKeysAndPartialAddress)]
+fn get_public_keys_and_partial_address_oracle(_address: AztecAddress) -> [Field; 9] {}
+
+unconstrained fn get_public_keys_and_partial_address_oracle_wrapper(address: AztecAddress) -> [Field; 9] {
+  get_public_keys_and_partial_address_oracle(address)
+}
+
+pub fn get_public_keys_and_partial_address(address: AztecAddress) -> (PartialAddress, [GrumpkinPoint; 4]) {
+  let result = get_public_keys_and_partial_address_oracle_wrapper(address);
+
+  let partial_address = PartialAddress::from_field(result[0]);
+  let nullifier_pub_key = GrumpkinPoint::new(result[1], result[2]);
+  let incoming_pub_key = GrumpkinPoint::new(result[3], result[4]);
+  let outgoing_pub_key = GrumpkinPoint::new(result[5], result[6]);
+  let tagging_pub_key = GrumpkinPoint::new(result[7], result[8]);
+
+  _check_public_key_validity_constrain_oracle(
+    address,
+    partial_address,
+    nullifier_pub_key,
+    incoming_pub_key,
+    outgoing_pub_key,
+    tagging_pub_key,
+  );
+
+  (partial_address, [nullifier_pub_key, incoming_pub_key, outgoing_pub_key, tagging_pub_key])
+}
+
+fn _check_public_key_validity_constrain_oracle(
```

Comment:
I think it would likely be easier to follow, if you have that the function is returning the address instead, and then constrain it in the caller. Would then have function that is essentially take the preimage and compue, which should be similar to what we have other places (as jan mentioned)